### DISCUSSION TITLE
[REACTFLOw] - How to trigger event when node or edge is added or removed?

### DISCUSSION THREAD

**Discussion by sujit-baniya** [xyflow/xyflow#3845] [Q&A]
I didn't find any event to detect if node or edge is added on panel. Please suggest how I could find the node edge if added or remove?

Issue:

Currently I using `instance.addNodes(newNode)` to add node. Right after the line I'm trying to assign `parentNode` (new node id) to rest of the nodes. But it complains `Parent node with <id> not defined`.

As a temporary solution, I've added a timeout. 

```js
const newNode = {
                id: getId(),
                type: "group",
                position: pos,
                style: {
                    height: el.clientHeight + 50,
                    width: el.clientWidth + 50,
                },
                data: {label: `group node`},
                zIndex: 0
            };
            flowInstance.addNodes(newNode)
            setTimeout(() => {
                flowInstance.setNodes((nds) => {
                    return nds.map(n => {
                        nodes.forEach(nt => {
                            if (nt.id === n.id) {
                                n.parentNode = newNode.id
                            }
                        })
                        return n
                    })
                });
            }, 300)
```

Please suggest.

**Comment by bcakmakoglu** [comment]
You can use `onNodesChange` and `onEdgesChange` to listen for changes that have the type `add` or `remove` which would indicate if a node or edge was added or removed.

```ts
const onNodesChange = (changes) => {
	const addChanges = changes.filter(change => change.type === 'add')
	const removeChanges = changes.filter(change => change.type === 'remove')

	// do something...
}

<ReactFlow nodes={nodes} edges={edges} onNodesChange={onNodesChange} />
```

Although don't forget to apply the changes to your nodes/edges after you went through them otherwise it'll seem like nothing is happening 😄

**Reply by sujit-baniya** [reply]
Thanks it worked on `<ReactFlow>` . Is it possible to get same by using hook `useReactFlow()`?
